core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -949,9 +953,65 @@ public void onSuccess() {
                                         }
 
                                         @Override
-                                        public void onFailure(Throwable t) {
-                                            // TODO: handle catastrophic non-channel failures
-                                            onReplicaFailure(nodeId, exp);
+                                        public void onFailure(Throwable shardFailedError) {
+                                            logger.error("[{}] catastrophic error while failing replica shard [{}] for [{}]", shardFailedError, shardId, shard, exp);
+                                            if (shardFailedError instanceof ShardStateAction.NoLongerPrimaryShardException) {
```

Comment:
@bleskes I'm not sure about this, can you try to convince me? In particular, this seems like a drastic course of action to take when we do not know what happened.